src/main/java/com/mixpanel/android/viewcrawler/ImageStore.java

```diff
@@ -0,0 +1,138 @@
+import java.io.OutputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+/**
+ * ABSOLUTELY NOT thread, or even process safe.
+ * Writes and reads files and directories at known paths, and uses a shared instance of MessageDigest.
+ */
+/* package */ class ImageStore {
+    public ImageStore(Context context) {
+        this(
+            context.getDir(DEFAULT_DIRECTORY_NAME, Context.MODE_PRIVATE),
+            new HttpService()
+        );
+    }
+
+    public ImageStore(File directory, RemoteService poster) {
+        mDirectory = directory;
+        mPoster = poster;
+        MessageDigest useDigest;
+        try {
+            useDigest = MessageDigest.getInstance("SHA1");
+        } catch (NoSuchAlgorithmException e) {
+            Log.w(LOGTAG, "Images won't be stored because this platform doesn't supply a SHA1 hash function");
+            useDigest = null;
+        }
+
+        mDigest = useDigest;
+    }
+
+    public Bitmap getImage(String url) {
+        final File file = storedFile(url);
+        byte[] bytes = null;
+        if (null == file || !file.exists()) {
+            try {
+                bytes = mPoster.performRequest(url, null);
+            } catch (IOException e) {
+                Log.i(LOGTAG, "Can't download bitmap", e);
+            } catch (RemoteService.ServiceUnavailableException e) {
+                Log.i(LOGTAG, "Couldn't download ");
+            }
+        }
+
+        final Bitmap bitmap;
+        if (null != bytes) {
+            if (null != file && bytes.length < MAX_BITMAP_SIZE) {
+                OutputStream out = null;
+                try {
+                    out = new FileOutputStream(file);
+                    out.write(bytes);
+                } catch (FileNotFoundException e) {
+                    Log.e(LOGTAG, "It appears that ImageStore is misconfigured, or disk storage is unavailable- can't write to bitmap directory", e);
+                } catch (IOException e) {
+                    Log.w(LOGTAG, "Can't store bitmap", e);
+                } finally {
+                    if (null != out) {
+                        try {
+                            out.close();
+                        } catch (IOException e) {
+                            Log.w(LOGTAG, "Problem closing output file", e);
+                        }
+                    }
+                }
+            }
+
+            bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.length);
+            if (null == bitmap) {
+                Log.w(LOGTAG, "Downloaded bitmap could not be interpreted");
+            }
+        } else {
+            bitmap = BitmapFactory.decodeFile(file.getAbsolutePath());
+            if (null == bitmap) {
+                Log.w(LOGTAG, "Bitmap on disk can't be opened or is corrupt");
+                final boolean ignored = file.delete();
+            }
+        }
+
+        return bitmap;
+    }
+
+    public void clearStorage() {
```

Comment:
now, the only question I have is - do we remove the old image once it gets replaced with a new image from memory? It may potentially eat up a lot of memory.